docs/design-proposals/multi-tenancy/multi-tenancy-schduler-agent-and-controllers.md

```diff
@@ -0,0 +1,69 @@
+# Scheduler, Node Agent and Controllers in Arktos Multi-Tenancy
+
+Xiaoning Ding, Qian Chen, Peng Du
+
+## Introduction
+
+This document is a part of the multi-tenancy design documents in Arktos. It describes most of the non-API-Server changes, including changes in scheduler, node agent and controllers.
+
+Please refer to other design documents in the "multi-tenancy" document directory if you want to learn about other parts in the multi-tenancy design.
+
+## Scheduler
+
+Scheduler and controllers work in a similar way. They all list/watch some API objects and then make update to API certain objects. This section discusses some common changes that are required in schedulers and controllers to make them tenancy-aware. And also a new tenant controller.
+
+### General Tenancy-aware changes
+
+Scheduler and controllers need to use right client-go APIs to list/watch all objects, or update an object in a certain world. Also world-matching is required in their logics.
+
+Take the scheduler as an example:
+
+* When list/watch objects like pods or pvs, it should list/watch all objects in system world and tenant worlds using credentials of system user. This is achieved by calling client-go API with TenantAll:
+
+```go
+Client.CoreV1().PodsWithMultiTenancy(pod.Namespace, metav1.TenantAll).List()
+```
+
+* When correlating a pod to other resources, world comparison needs to be added:
+
+```go
+(... existing comparisons ) && (pod.Tenant == rs.Tenant)
+```
+ 
+* When update a certain pod, it should use world-specific calls: 
+
+```go
+Client.CoreV1().PodsWithMultiTenancy(pod.Namespace, pod.Tenant).UpdateStatus(pod)
+```
+
+### P2: Tenancy-Fairness
+
+Tenancy-faireness is required in schedulers and most controllers. It ensures one tenant won't cause unresponsiveness of other tenants. For example, scheduler shouldn't only schedule pods from one tenant in a certain period. This would make other tenants feel that scheduling is stopped in their world.
+
+The implementation varies depending on the logics of different controllers.
+
+## Controllers
+
+### New: Tenant Controller
+
+A new controller, tenant controller, is introduced to reconcile the changes related to API resource /api/v1/tenants/{tenantName}.
+
+When a new tenant object is created, tenant controller creates the following default resources for the new tenant:
+
+   * /api/v1/tenants/{tenantName}/namespaces/default
+   * /api/v1/tenants/{tenantName}/namespaces/kube-system
+   * /api/v1/tenants/{tenantName}/namespaces/kube-public
+   * /api/v1/tenants/{tenantName}/networks/default
+ 
+The next-level default resources, such as such as the default service account for a namespace and default DNS service for a network, will be created by the corresponding controllers.
```

Comment:
What is the behavior when a tenant is deleted? Namespace controller will delete all the resources in the namespace when the namespace is deleted. I think we can do it in two phases: Ph-1 delete all the resources in this tenant, PH-2, save a copy of the data for a given period time before a complete deletion.

Reply:
such as such as

Reply:
> such as such as

Good catch. Will fix.

Reply:
> What is the behavior when a tenant is deleted? Namespace controller will delete all the resources in the namespace when the namespace is deleted. I think we can do it in two phases: Ph-1 delete all the resources in this tenant, PH-2, save a copy of the data for a given period time before a complete deletion.

Agree on the suggestion and phase-1.  

Not sure if we need to do phase-2. It's different than deleting a tenant from IAM.  We can add an issue to track phase-2 in case there are requirements but I don't suggest add phase-2 to the design for now.

Reply:
Updated.